src/main/java/net/snowflake/client/jdbc/ConnectionHandler.java

```diff
@@ -0,0 +1,42 @@
+package net.snowflake.client.jdbc;
+
+import java.sql.*;
+import java.util.Properties;
+import net.snowflake.client.core.BaseSFSession;
+import net.snowflake.client.core.SFBaseResultSet;
+
+/**
+ * Factory class that presents the implementation of a Snowflake Connection. This allows for
+ * alternate definitions of SFSession, SFStatement, and SFResultSet, (representing the 'physical'
+ * implementation layer) that can share high-level code.
+ */
+public interface ConnectionHandler {
+
+  /**
+   * Whether this Connection supports asynchronous queries. If yes, createAsyncResultSet may be
+   * called.
+   */
+  boolean supportsAsyncQuery();
+
+  /** Initializes the SnowflakeConnection */
+  void initializeConnection(String url, Properties info) throws SQLException;
+
+  /** Gets the SFSessionInterface implementation for this connection implementation */
+  BaseSFSession getSFSession();
+
+  /** Creates a result set from a query id. */
+  ResultSet createResultSet(String queryID, Connection connection) throws SQLException;
```

Comment:
Why pass connection here ? I think `Statement` would be cleaner.

Reply:
Good suggestion. I'll change this in the next PR, when the `Statement` interface is introduced. That will make the change cleaner.